docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,795 @@
+until it has discovered an elected master node. If no elected master is
+discovered then the node will retry after `discovery.find_peers_interval` which
+defaults to `1s`.
+
+If the node is master-eligible then it continues this discovery process until it
+has either discovered an elected master node or else it has discovered enough
+masterless master-eligible nodes to complete an election. If neither of these
+occur quickly enough then the node will retry after
+`discovery.find_peers_interval` which defaults to `1s`.
+
+The discovery process is controlled by the following settings.
+
+`discovery.find_peers_interval`::
+
+    Sets how long a node will wait before attempting another discovery round.
+    Defaults to `1s`.
+
+`discovery.request_peers_timeout`::
+
+    Sets how long a node will wait after asking its peers again before
+    considering the request to have failed. Defaults to `3s`.
+
+`discovery.probe.connect_timeout`::
+
+    Sets how long to wait when attempting to connect to each address. Defaults
+    to `3s`.
+
+`discovery.probe.handshake_timeout`::
+
+    Sets how long to wait when attempting to identify the remote node via a
+    handshake. Defaults to `1s`.
+
+`discovery.cluster_formation_warning_timeout`::
+
+    Sets how long a node will try to form a cluster before logging a warning
+    that the cluster did not form. Defaults to `10s`.
+
+If a cluster has not formed after `discovery.cluster_formation_warning_timeout`
+has elapsed then the node will log a warning message that starts with the phrase
+`master not discovered` which describes the current state of the discovery
+process.
+
+[float]
+[[modules-discovery-bootstrap-cluster]]
+=== Bootstrapping a cluster
+
+Starting an Elasticsearch cluster for the very first time requires the initial
+set of master-eligible nodes to be explicitly set on one or more of the
+master-eligible nodes in the cluster using this setting:
+
+`cluster.initial_master_nodes`::
+
+    Sets a list of the <<node.name,node names>> or transport addresses of the
+    initial set of master-eligible nodes in a brand-new cluster. By default
+    this list is empty, meaning that this node expects to join a cluster that
+    has already been bootstrapped.
+
+This setting can be given on the command line when starting up each node, or
+added to the `elasticsearch.yml` configuration file. Once the cluster has
+formed this setting is no longer required and should be removed.
+
+For a cluster with 3 master-eligible nodes (with <<node.name,node names>>
+`master-a`, `master-b` and `master-c`) the configuration will look as follows:
+
+[source,yaml]
+--------------------------------------------------
+cluster.initial_master_nodes:
+  - master-a
+  - master-b
+  - master-c
+--------------------------------------------------
+
+Alternatively the IP addresses or hostnames
+(<<node.name,if node name defaults to the host name>>) can be used. If there
```

Comment:
Hostnames (from `publish_host`) are accepted _as well as_ node names aren't they?

Reply:
No, for now `matchesRequirement` does not check that AFAICS.